src/command/run.go

```diff
@@ -59,15 +71,35 @@ func RunWith(opts Options, cmd string, args ...string) (*Result, error) {
 	if opts.Env != nil {
 		subProcess.Env = opts.Env
 	}
-	output, err := subProcess.CombinedOutput()
+	var output bytes.Buffer
+	subProcess.Stdout = &output
+	subProcess.Stderr = &output
+	result := Result{command: cmd, args: args}
+	input, err := subProcess.StdinPipe()
+	if err != nil {
+		return &result, err
+	}
+	err = subProcess.Start()
+	if err != nil {
+		return &result, fmt.Errorf("can't start subprocess '%s %s': %w", cmd, strings.Join(args, " "), err)
+	}
+	for _, userInput := range opts.Input {
+		// Here we simply wait for some time until the subProcess needs the input.
+		// Capturing the output and scanning for the actual content needed
+		// would introduce substantial amounts of multi-threaded complexity
+		// for not enough gains.
```

Comment:
I had implemented scanning the output stream of the subshell. While the code was relatively straightforward, I ran into  timing and concurrency issues. This implementation is much simpler and more robust, and shouldn't be much slower.

Reply:
Can you explain the issues you hit? I don't think waiting for some delay is more robust

Reply:
It was some type of Heisenbug. My code was reading the stream just fine, in many chunks. All the unit tests to verify the stream reading in isolation passed. But under load, it sometimes didn't receive the last chunk of output. Instead, it received an error that the pipe has already been closed. Apparently, reading from a closed pipe is not allowed in Go, even if there was still unread content in it. 

I don't know exactly why this problem happened. Given that this problem reliably doesn't happen with a `bytes.Buffer`, I assume it could be some sort of timing issue. My reader doesn't just read from the pipe, it also signals via a separate channel that new output has been received (so that it can be scanned for matches). My current hypothesis is that this notification takes too long (since writing to pipes is synchronized), and then the running process ends before my reader can consume all of the output.

After spending two days investigating this, I gave up since imo it's not worth spending so much time on such a peripheral issue when there are workarounds like the one implemented. 

Scanning the output stream for text introduces a good amount of additional infrastructure and complexity. For example, a number of new structs, functions, and channels to accumulate the stdout text and scan it for content. The current implementation is much simpler and a lot more straightforward. We use this only two dozen times in the entire test suite. So, this only saves us 1-2 seconds or 0.5-1% of the total time to run all cukes. Optimizing this costs a lot and doesn't move the needle. As long as this runs reliably on TravisCI (which is really slow), it's good enough for me.

What do you think?

Reply:
Knew this felt familiar. Were you using this: https://github.com/Originate/go-execplus

I'm fine as is but if didn't try go-execplus can you add to the comment that we could try it?

Reply:
Ha, no I wasn't aware that go-execplus exists! I added a TODO comment that mentions it. I'm happy to try using it later, when I get my wind back to debug more multi-threaded concurrency issues. Did you run go-execplus at scale? The problem I encountered only happens under concurrent load. Running individual subprocesses was fine.